### PR TITLE
feat: refinance sources in batch

### PR DESCRIPTION
**Pull request by mishuagopian**

### CODE REVIEW

File: src/contracts/MslV4.ts [hunk 1]
```diff
@@ -236,17 +236,51 @@ export class MslV4 extends BaseContract<typeof multiSourceLoanABIV4> {
...
+    // Generate (renegotiation offer, loan) pairs
+    refinancings.forEach(({ loan, sources, newAprBps }, index) => {
+      const targetPrincipal = loan.source.map(({ principalAmount, loanId }) => {
+        const refinancingSource = sources.find(({ source }) => source.loanId === loanId);
+        return refinancingSource ? principalAmount - refinancingSource.refinancingPrincipal : 0n;
+      });
+      const refinancingPrincipalAmount = sources.reduce(
+        (acc, { refinancingPrincipal }) => acc + refinancingPrincipal,
+        0n,
+      );
```
Comment (franco-gondi) [inline src/contracts/MslV4.ts:265]: I would move that sum to a helper function
```suggestion
      const refinancingPrincipalAmount = sum(sources.map({refinancingPrincipal} => refinancingPrincipal))
```

```suggestion
      const refinancingPrincipalAmount = sumBy(sources, 'refinancingPrincipal')
```

File: src/contracts/MslV4.ts [hunk 2]
```diff
@@ -236,17 +236,53 @@ export class MslV4 extends BaseContract<typeof multiSourceLoanABIV4> {
...
+    const offers: RenegotiationV4[] = [];
+    const loans: LoanV4[] = [];
+
+    // Generate (renegotiation offer, loan) pairs
+    refinancings.forEach(({ loan, sources, newAprBps }, index) => {
+      const targetPrincipal = loan.source.map(({ principalAmount, loanId }) => {
+        const refinancingSource = sources.find(({ source }) => source.loanId === loanId);
+        return refinancingSource
+          ? principalAmount - refinancingSource.refinancingPrincipal
+          : principalAmount;
```
Comment (franco-gondi) [inline src/contracts/MslV4.ts:262]:
```suggestion
        return principalAmount - (refinancingSource?.refinancingPrincipal ?? 0)
```

File: src/contracts/MslV5.ts [hunk 3]
```diff
@@ -310,17 +310,61 @@ export class MslV5 extends BaseContract<typeof multiSourceLoanABIV5> {
...
+        renegotiationId: renegotiationId + BigInt(index),
+        loanId: loan.source[0].loanId,
+        lender: this.wallet.account.address,
+        fee: 0n,
+        targetPrincipal,
+        principalAmount: refinancingPrincipalAmount,
+        aprBps: newAprBps,
+        expirationTime: 0n,
+        duration: 0n,
+      };
```
Comment (franco-gondi) [inline src/contracts/MslV5.ts:350]: this is repeated

Comment (franco-gondi) [reply src/contracts/MslV5.ts:350]: offer generation is the same in v4 and v5, the only thing different is the multicall.
I would probably abstract one of the two

File: src/contracts/MslV6.ts [hunk 4]
```diff
@@ -356,6 +356,73 @@ export class MslV6 extends BaseContract<typeof multiSourceLoanAbiV6> {
...
+      const trancheIndex = sources.map(({ source }) => BigInt(source.loanIndex));
+
+      const refinancingPrincipalAmount = sources.reduce(
+        (acc, { refinancingPrincipal }) => acc + refinancingPrincipal,
+        0n,
+      );
+
+      const offer = {
+        renegotiationId: renegotiationId + BigInt(index),
+        loanId: BigInt(Math.max(...loan.tranche.map(({ loanId }) => Number(loanId)))),
```
Comment (franco-gondi) [inline src/contracts/MslV6.ts:384]: I would move this to a loan helper

File: src/contracts/MslV4.ts [hunk 5]
```diff
@@ -236,17 +236,53 @@ export class MslV4 extends BaseContract<typeof multiSourceLoanABIV4> {
...
+      });
+      const refinancingPrincipalAmount = sources.reduce(
+        (acc, { refinancingPrincipal }) => acc + refinancingPrincipal,
+        0n,
+      );
+
+      offers.push({
+        renegotiationId: renegotiationId + BigInt(index),
+        loanId: loan.source[0].loanId,
```
Comment (franco-gondi) [inline src/contracts/MslV4.ts:271]: I would move this to a loan helper

Comment (franco-gondi) [reply src/contracts/MslV4.ts:271]: something like getCurrentContractLoanId

File: src/gondi.ts [hunk 6]
```diff
@@ -511,6 +516,102 @@ export class Gondi {
...
+      const currentLoanRefinancings = refisByContract[version][tokenLoanId];
+
+      const currentSourceNewAprBps = BigInt(
+        Math.floor(Number(source.aprBps) * (1 - aprBpsImprovementPercentage)),
+      );
+      const newAprBps =
+        currentLoanRefinancings?.newAprBps &&
+        currentLoanRefinancings?.newAprBps < currentSourceNewAprBps
+          ? currentLoanRefinancings.newAprBps
+          : currentSourceNewAprBps;
```
Comment (franco-gondi) [inline src/gondi.ts:584]:
```suggestion
      const newAprBps =
        currentLoanRefinancings?.newAprBps &&
        min(currentLoanRefinancings?.newAprBps, currentSourceNewAprBps)
```

Comment (franco-gondi) [inline src/gondi.ts:584]:
```suggestion
      const newAprBps = minOfDefined(currentLoanRefinancings?.newAprBps, currentSourceNewAprBps)
```

File: src/contracts/MslV4.ts (hunk 1 above)
Comment (mishuagopian) [reply src/contracts/MslV4.ts]: https://github.com/gondixyz/gondi-js/pull/65/commits/e347f5f9a259169c48335a3a1dc4d4dcc0121816

File: src/contracts/MslV4.ts (hunk 2 above)
Comment (mishuagopian) [reply src/contracts/MslV4.ts]: https://github.com/gondixyz/gondi-js/pull/65/commits/82fb04f6f8695a2d39e7c488b033161f1b8297d8

File: src/contracts/MslV6.ts (hunk 4 above)
Comment (mishuagopian) [reply src/contracts/MslV6.ts]: https://github.com/gondixyz/gondi-js/pull/65/commits/877cd89275c0bdbff12bd5e4756caf73ec9110c4

File: src/gondi.ts (hunk 6 above)
Comment (mishuagopian) [reply src/gondi.ts]: https://github.com/gondixyz/gondi-js/pull/65/commits/c728b8e293042747808e27fafd6774b40fe429cf

File: src/contracts/MslV5.ts (hunk 3 above)
Comment (mishuagopian) [reply src/contracts/MslV5.ts]: Let's do it in another PR.

File: src/contracts/MslV6.ts [hunk 7]
```diff
@@ -356,6 +356,73 @@ export class MslV6 extends BaseContract<typeof multiSourceLoanAbiV6> {
...
+          hash: txHash,
+        });
+        const filter = await this.contract.createEventFilter.LoanRefinanced();
+        const events = filterLogs(receipt, filter);
+        if (events.length !== refinancings.length) throw new Error('Loan not refinanced');
+
+        const results = events.map(({ args }) => args);
+        return { results, ...receipt };
+      },
+    };
```
[the remark this answers is not in the document]
Comment (mishuagopian) [reply src/contracts/MslV6.ts:420]: Let's do it in another PR.

**Review by franco-gondi — APPROVED** [review]

**Pull request merged**